src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,393 @@
+        syntaxNode switch
+        {
+            LocalDeclarationStatementSyntax @localDeclaration => @localDeclaration.Declaration,
+            ExpressionStatementSyntax @expressionStatement => @expressionStatement.Expression,
+            _ => syntaxNode
+        };
+
+    private static void RemoveVariablesFromContext(IEnumerable<string> variablesToRemove, ContextInformation contextInformation)
+    {
+        var variableValues = contextInformation.VariableValues;
+        var expressionAnalysisContexts = contextInformation.ExpressionAnalysisContexts;
+
+        foreach (var variable in variablesToRemove)
+        {
+            if (!variableValues.ContainsKey(variable))
+            {
+                continue;
+            }
+
+            var previousValue = variableValues[variable];
+            var locationsArray = previousValue.Locations.ToArray();
+            var previousContext = previousValue.ExpressionAnalysisContext with { Node = previousValue.ExpressionAnalysisContext.Node with { Locations = locationsArray } };
+            expressionAnalysisContexts.Add(previousContext);
+        }
+
+        variableValues.RemoveRange(variablesToRemove);
+    }
+
+    private static void ProcessBuildersExpression(SyntaxNode syntaxNode, ContextInformation contextInformation, HashSet<string> dirtyVariables)
+    {
+        syntaxNode = GetUnderlyingExpression(syntaxNode);
+
+        if (syntaxNode is VariableDeclarationSyntax variableDeclaration)
+        {
+            foreach (var declaration in variableDeclaration.Variables)
+            {
+                ProcessDeclaration(declaration, contextInformation, dirtyVariables);
+            }
+        }
+        else if (syntaxNode is AssignmentExpressionSyntax assignmentExpression)
+        {
+            ProcessAssignment(assignmentExpression, contextInformation, dirtyVariables);
+        }
+    }
+    
+    private static bool IsValidBuildersExpression(SyntaxNode syntaxNode, ContextInformation contextInformation)
+    {
+        var isBuilders = false;
+        var semanticModel = contextInformation.SemanticModel;
+        syntaxNode = GetUnderlyingExpression(syntaxNode);
+
+        if (syntaxNode is VariableDeclarationSyntax variableDeclaration)
+        {
+            var symbol = semanticModel.GetSymbolInfo(variableDeclaration.Type).Symbol as ITypeSymbol;
+            isBuilders = symbol.IsBuilderDefinition();
+        }
+        else if (syntaxNode is AssignmentExpressionSyntax assignmentExpression)
+        {
+            var variableIdentifier = assignmentExpression.Left.DescendantNodesAndSelf().OfType<IdentifierNameSyntax>().FirstOrDefault();
+            var symbol = semanticModel.GetTypeInfo(variableIdentifier).Type;
+            isBuilders = symbol.IsBuilderDefinition();
+        }
+
+        return isBuilders;
+    }
+
+    private static bool IsConditionalBlock(SyntaxNode syntaxNode) =>
+        syntaxNode switch
+        {
+            SwitchStatementSyntax or
```

Comment:
Probably no need for SwitchStatementSyntax and SwitchExpressionSyntax, IfStatementSyntax, TryStatementSyntax ? Need to double check.

Reply:
Done